source/grid/tria_description.cc

```diff
@@ -0,0 +1,662 @@
+            if (cell->is_active() && cell->subdomain_id() == my_rank)
+              add_vertices_of_cell_to_vertices_owned_by_locally_owned_cells(
+                cell, vertices_owned_by_locally_owned_cells);
+
+          // helper function to determine if cell is locally relevant
+          // (i.e. a cell which is connected to a vertex via a locally owned
+          // active cell)
+          auto is_locally_relevant = [&](
+                                       TriaIterator<CellAccessor<dim, spacedim>>
+                                         &cell) {
+            for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+                 ++v)
+              if (vertices_owned_by_locally_owned_cells[cell->vertex_index(v)])
+                return true;
+            return false;
+          };
+
+          // 2) process all local and ghost cells: set up needed data
+          // structures and collect all locally relevant vertices
+          // for second sweep
+          std::map<unsigned int, unsigned int> vertices_locally_relevant;
+          construction_data.cell_infos.resize(1);
+
+          for (auto cell : tria.cell_iterators())
+            if (cell->is_active() && is_locally_relevant(cell))
+              {
+                // to be filled
+                CellData<dim> cell_info;
+
+                // a) extract cell definition (with old numbering of
+                // vertices)
+                dealii::CellData<dim> cell_data;
+                cell_data.material_id = cell->material_id();
+                cell_data.manifold_id = cell->manifold_id();
+                for (unsigned int v = 0;
+                     v < GeometryInfo<dim>::vertices_per_cell;
+                     ++v)
+                  cell_data.vertices[v] = cell->vertex_index(v);
+                construction_data.coarse_cells.push_back(cell_data);
+
+                // b) save indices of each vertex of this cell
+                for (unsigned int v = 0;
+                     v < GeometryInfo<dim>::vertices_per_cell;
+                     ++v)
+                  vertices_locally_relevant[cell->vertex_index(v)] =
+                    numbers::invalid_unsigned_int;
+
+                // c) save boundary_ids of each face of this cell
+                for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell;
+                     ++f)
+                  {
+                    types::boundary_id boundary_ind =
+                      cell->face(f)->boundary_id();
+                    if (boundary_ind != numbers::internal_face_boundary_id)
+                      cell_info.boundary_ids.emplace_back(f, boundary_ind);
+                  }
+
+                // d) compute a new coarse-cell id (by ignoring all parent
+                // level)
+                types::coarse_cell_id new_coarse_cell_id =
+                  convert_cell_id_binary_type_to_level_coarse_cell_id<dim>(
+                    cell->id().template to_binary<dim>());
```

Comment:
Please move the functionality you have in `convert_cell_id_binary_type_to_level_coarse_cell_id` to a member function of `CellId`. What you are doing here is first convert to a binary representation, and then using special knowledge of what's in that binary representation to extract information from the original `CellId`. That's backward and also dangerous because you assume information you shouldn't have. Just let the `CellId` class do this.

Reply:
As discussed in PR #9391, I would like to construct per default all multigrid levels and remove the flattening function here. As a consequence, this function will also not be called anymore and can be removed.

Reply:
OK, I look forward to the function disappearing. Please open a github issue that reminds us before the next release to make sure the function is indeed gone!